### PR TITLE
Listenable app storage

### PR DESCRIPTION
**Pull request by geofjamg**

### CODE REVIEW

File: afs/afs-core/src/main/java/com/powsybl/afs/ProjectFileCreationContext.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        this.fileSystem = Objects.requireNonNull(fileSystem);
+    }
+
+    private final NodeInfo info;
+
+    private final ListenableAppStorage storage;
+
+    private final NodeInfo projectInfo;
+
+    private final AppFileSystem fileSystem;
```
Comment (mathbagu) [inline afs/afs-core/src/main/java/com/powsybl/afs/ProjectFileCreationContext.java]: Why members are declared after constructor? This class is similar to ProjectFileBuildContext (refactor?)

File: afs/afs-storage-api/src/main/java/com/powsybl/afs/storage/AppStorageListener.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public interface AppStorageListener {
+
+    void nodeCreated(NodeId id);
+
+    void nodeRemoved(NodeId id);
+
+    void attributeUpdated(NodeId id, String attributeName);
```
Comment (mathbagu) [inline afs/afs-storage-api/src/main/java/com/powsybl/afs/storage/AppStorageListener.java]: Do we need to have the old and the value here?

File: afs/afs-storage-api/src/main/java/com/powsybl/afs/storage/DefaultListenableAppStorage.java [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+import java.util.*;
+
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class DefaultListenableAppStorage implements ListenableAppStorage {
+
+    private final AppStorage storage;
+
+    private final WeakHashMap<Object, AppStorageListener> listeners = new WeakHashMap<>();
```
Comment (mathbagu) [inline afs/afs-storage-api/src/main/java/com/powsybl/afs/storage/DefaultListenableAppStorage.java]: Only one listener per Object? Consider using a multimap here?

File: afs/afs-core/src/main/java/com/powsybl/afs/ProjectFileCreationContext.java (hunk 1 above)
Comment (geofjamg) [reply afs/afs-core/src/main/java/com/powsybl/afs/ProjectFileCreationContext.java]: Fixed for 1er question + common context base class added

File: afs/afs-storage-api/src/main/java/com/powsybl/afs/storage/AppStorageListener.java (hunk 2 above)
Comment (geofjamg) [reply afs/afs-storage-api/src/main/java/com/powsybl/afs/storage/AppStorageListener.java]: Maybe later on in a future version

File: afs/afs-storage-api/src/main/java/com/powsybl/afs/storage/DefaultListenableAppStorage.java (hunk 3 above)
Comment (geofjamg) [reply afs/afs-storage-api/src/main/java/com/powsybl/afs/storage/DefaultListenableAppStorage.java]: Fixed

**Pull request merged**